src/middlewared/middlewared/plugins/vm.py

```diff
@@ -605,6 +674,55 @@ class RAW(StorageDevice):
     )
 
 
+class PCI(Device):
+
+    schema = Dict(
+        'attributes',
+        Str('pptdev', default=None, required=True),
+    )
+
+    def __init__(self, data, middleware=None):
+        Device.__init__(self, data, middleware)
+        self.ppt_map = {'host_bsf': None, 'guest_bsf': None}
+
+    def xml(self, *args, **kwargs):
+        host_bsf = kwargs.pop('host_bsf')
+        guest_bsf = kwargs.pop('guest_bsf')
+
+        self.ppt_map = {'host_bsf': host_bsf, 'guest_bsf': guest_bsf}
+        return create_element(
```

Comment:
If I understand correctly, this is not required right now as PCI currently works with bhyve args with the libvirt version we have. How about we use the `LIBVIRT_HOSTDEV` variable here as well to make sure we don't generate this and add it to the xml even though it won't work and we have to make it work via command line args ?

Reply:
Maybe ok now since the fbuf/vncserver is passed with command-line arg. Before, ordering of devices was not consistent due to fbuf/vncserver not getting a fixed PCI address. I think previously there was also a USB controller 'floating around'; not anymore. Those devices caused PCI address conflicts. Inclusion of the xml also when using bhyve args avoided the conflicts as libvirt seemed to avoid the PCI address provided in the xml for the passthru device when libvirt assigned addresses/slots for other devices.

I have made xml generation conditioned on LIBVIRT_HOSTDEV, but added a comment about potential address conflicts, in case the issue reappears.

Reply:
Thank you for elaborating @stattin42, I see what the issue is. So what happens is that libvirt chooses slots for the controller itself and earlier it was doing that for VNC too. Now moving on, libvirt is still handling that for `tablet`, however I have created a PR to address that ( https://github.com/freenas/freenas/pull/4549 ). So we shouldn't be facing that issue anymore. 

I think it's safe to say that we are good for this. Can you please rebase and test to be certain ?

Reply:
@stattin42 I have been going through libvirt source and it is going to keep on adding an extra controller each time we cross the AHCI devices allowed in one slot which means there's potential for conflict. Unfortunately there doesn't seem to be a good way to resolve this except for patching bhyve driver to do this properly as it's just consuming an entire slot. So I think the best case for us is to keep on using xml arguments as well for passthrough.

Also can we remove `LIBVIRT_HOSTDEV` variable references ? When libvirt bhyve driver has support for it, we will not be keeping backward compatibility like this, we will just remove the bhyve args code ; ) Older version releases use a different stable branch so I don't anticipate a scenario where this variable and it's usages will help us. Please use tomorrow nightlies for testing as we are updating ports tree and we will have a newer version of libvirt in to work with.

Reply:
Ok. Thanks for checking @sonicaj. I just tested with PR #4549. Worked with my test configurations (with only two ahci devices). Will update though, based on your findings for more complex configurations, to always include xml for passthrough. I will remove the LIBVIRT_HOSTDEV variable too, but keep the comments with reasons for using both xml and bhyve args.